mods/cnc/maps/twist-of-fate/twist-of-fate-AI.lua

```diff
@@ -0,0 +1,647 @@
+			Trigger.OnKilled(newBuilding, function()
+				ProductionQueue["vehicle"] = { }
+				CheckTeamCompleted()
+				WaitingAirDrop = false
+			end)
+			RestartUnitProduction()
+			NeedHarv = false
+		elseif newBuilding.Type == "hpad" then
+			ProductionBuildings["aircraft"] = newBuilding
+			Trigger.OnKilled(newBuilding, function()
+				ProductionQueue["aircraft"] = { }
+				CheckTeamCompleted()
+			end)
+			RestartUnitProduction()
+		end
+
+		Trigger.AfterDelay(50, function() CheckBase() end)
+	end)
+
+end
+
+-- Units production logic
+UniqueTeamsQueue = { }
+ProductionCooldown = false
+CheckProduction = function()
+	if Utils.All(ProductionBuildings, function(b) return b.IsDead end) then
+		return
+	elseif #Nod.GetActorsByType("proc") < 1 and Nod.Resources < 6000 then
+		RestartUnitProduction()
+		return
+	elseif not ProductionBuildings["vehicle"].IsDead and not ProductionBuildings["vehicle"].IsProducing("harv") and CheckForHarvester() then
+		NeedHarv = true
+		WaitingAirDrop = true
+		ProductionBuildings["vehicle"].Build({ "harv" }, function()
+			Trigger.AfterDelay(2, function()
+				WaitingAirDrop = false
+			end)
+			CheckProduction()
+			if Utils.RandomInteger(0,2) == 1 then
+				Trigger.AfterDelay(DateTime.Seconds(5), function()
+					if not ProductionBuildings["vehicle"].IsDead then
+						ProductionBuildings["vehicle"].Build({ "harv" })
+					end
+				end)
+			end
+		end)
+		RestartUnitProduction()
+		return
+	end
+	NeedHarv = false
+
+	if Producing and #ProductionQueue["infantry"] + #ProductionQueue["vehicle"] + #ProductionQueue["aircraft"] < 1 then
+		Producing = false
+	end
+	if ProductionCooldown and #UniqueTeamsQueue < 1 or Nod.Resources < 4000 then
+		RestartUnitProduction()
+		return false
+	else
+		ProduceUnits()
+		return true
+	end
+end
+
+ReduceProdCD = function()
+	Trigger.AfterDelay(DateTime.Minutes(2), function()
+		ProductionCooldownSeconds = ProductionCooldownSeconds - 1
+		if ProductionCooldownSeconds > 10 then
+			ReduceProdCD()
+		end
+	end)
+end
+
+ProduceUnits = function()
+	if NeedHarv then
+		RestartUnitProduction()
+		return
+	end
+	if not Producing then
+		NewTeam = { }
+		CreateUnitsGroup()
+		if #ProductionQueue["infantry"] + #ProductionQueue["vehicle"] + #ProductionQueue["aircraft"] > 0 then
+			Producing = true
+		else
+			RestartUnitProduction()
+			return
+		end
+	end
+
+	if #ProductionQueue["infantry"] > 0 and not ProductionBuildings["infantry"].IsDead and not ProductionBuildings["infantry"].IsProducing("e1") then
+		ProductionBuildings["infantry"].Build({ ProductionQueue["infantry"][1] }, function(u)
+			table.insert(NewTeam, u[1])
+			table.remove(ProductionQueue["infantry"], 1)
+			CheckTeamCompleted()
+		end)
+	end
+	if #ProductionQueue["vehicle"] > 0 and not ProductionBuildings["vehicle"].IsDead and not ProductionBuildings["vehicle"].IsProducing("harv") and not WaitingAirDrop then
+		ProductionBuildings["vehicle"].Build({ ProductionQueue["vehicle"][1] }, function(u)
+			table.insert(NewTeam, u[1])
```

Comment:
I just got a weird crash on my latest play through:
<details>

```
OpenRA engine version {DEV_VERSION}
OpenRA Language: en
Tiberian Dawn mod version {DEV_VERSION}
on map 5c165ed48eac6ce4b4288262f8df3304ee176ea7 (GDI - Twist of Fate by Westwood Studios).
Date: 2023-05-18 19:53:44Z
Operating System: Windows (X64, Microsoft Windows NT 10.0.19045.0)
Runtime Version: .NET CLR 6.0.16
Installed Language: en (Installed) en (Current) en (Current UI)
Exception of type `Eluant.LuaException`: [string "-- Make copies of global functions to preve..."]:30: Uncaught CLR exception at Lua->CLR boundary: System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
 ---> System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
 ---> Eluant.LuaException: [string "-- Make copies of global functions to preve..."]:30: Actor 'harv' does not define a property 'AttackMove'
   at Eluant.LuaRuntime.Call(IList`1 args)
   at Eluant.LuaRuntime.Call(LuaFunction fn, IList`1 args)
   at OpenRA.Mods.Common.Scripting.UtilsGlobal.Do(LuaValue[] collection, LuaFunction func) in \OpenRA\OpenRA.Mods.Common\Scripting\Global\UtilsGlobal.cs:line 29
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Span`1& arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at OpenRA.Scripting.ScriptMemberWrapper.Invoke(LuaVararg args) in \OpenRA\OpenRA.Game\Scripting\ScriptMemberWrapper.cs:line 76
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Span`1& arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at Eluant.LuaRuntime.MethodWrapper.Invoke(Object[] parms)
   at Eluant.LuaRuntime.MakeManagedCall(IntPtr state, MethodWrapper wrapper)
   at Eluant.LuaRuntime.Call(IList`1 args)
   at Eluant.LuaRuntime.Call(LuaFunction fn, IList`1 args)
   at OpenRA.Mods.Common.Scripting.ProductionQueueProperties.<>c__DisplayClass3_2.<Build>b__4(Actor factory, Actor unit) in \OpenRA\OpenRA.Mods.Common\Scripting\Properties\ProductionProperties.cs:line 172
   at OpenRA.Mods.Common.Scripting.ScriptTriggers.OpenRA.Mods.Common.Traits.INotifyProduction.UnitProduced(Actor self, Actor other, CPos exit) in \OpenRA\OpenRA.Mods.Common\Scripting\ScriptTriggers.cs:line 177
   at OpenRA.Mods.Common.Traits.Production.<>c__DisplayClass6_0.<DoProduction>b__0(World w) in \OpenRA\OpenRA.Mods.Common\Traits\Production.cs:line 99
   at OpenRA.World.Tick() in \OpenRA\OpenRA.Game\World.cs:line 453
   at OpenRA.Game.InnerLogicTick(OrderManager orderManager) in \OpenRA\OpenRA.Game\Game.cs:line 630
   at OpenRA.Game.LogicTick() in \OpenRA\OpenRA.Game\Game.cs:line 645
   at OpenRA.Game.Loop() in \OpenRA\OpenRA.Game\Game.cs:line 814
   at OpenRA.Game.Run() in \OpenRA\OpenRA.Game\Game.cs:line 867
   at OpenRA.Game.InitializeAndRun(String[] args) in \OpenRA\OpenRA.Game\Game.cs:line 305
   at OpenRA.Launcher.Program.Main(String[] args) in \OpenRA\OpenRA.Launcher\Program.cs:line 32
```
</details>

The Lua `Build` function has always been a bit buggy, so it's not out of the question that it somehow built a harvester and ran into the callback for normal vehicle production. (I don't know how else to explain a Harvester trying to attack move. Or do you have another idea?) Can we add a guard against that here? Something like
```suggestion
			if u[1].Type == "harv" then
				return
			end

			table.insert(NewTeam, u[1])
```

Reply:
That's why i added a delay between nod vehicles production. Not sure how to explain. The thing is... if one unit is ready, and a second one is queued before the plane arrives and leaves it on the floor, weird things happens... It's like the first unit doesn't execute his function and the next unit executes both, in this case the harvester try to attackmove.